contrib_bots/lib/xkcd.py

```diff
@@ -0,0 +1,139 @@
+from random import randint
+
+import logging
+import requests
+
+XKCD_TEMPLATE_URL = 'http://xkcd.com/%s/info.0.json'
+LATEST_XKCD_URL = 'http://xkcd.com/info.0.json'
+
+
+class XkcdHandler(object):
+    '''
+    This plugin provides several command that can be used for fetch comics from
+    xkcd. The bot looks for messages starting with "@xkcd" and responds with a
+    message with the comic based on provided commands.
+    '''
+
+    def usage(self):
+        return '''
+            This plugin allows users to fetch comics provided by xkcd.
+            Users should preface the command with "@xkcd".
+
+            There are several commands to use this bot:
+            - @xkcd help -> To show all commands the bot supports.
+            - @xkcd latest -> To fetch latest comic from xkcd.
+            - @xkcd random -> To fetch random comic from xkcd.
+            - @xkcd <comic_id> -> To fetch comic based on `<comic_id>`, e.g
+              `@xkcd 1234`.
+            '''
+
+    def triage_message(self, message, client):
+        # Return True if we want to (possibly) response to this message
+        original_content = message['content']
+        is_xkcd_called = original_content.startswith('@xkcd ')
+        is_xkcd_called_without_command = original_content == '@xkcd'
+
+        return is_xkcd_called or is_xkcd_called_without_command
+
+    def handle_message(self, message, client, state_handler):
+        xkcd_bot_response = get_xkcd_bot_response(message)
+
+        client.send_message(dict(
+            type='stream',
+            to=message['display_recipient'],
+            subject=message['subject'],
+            content=xkcd_bot_response,
+        ))
+
+
+class XkcdBotCommand(object):
+    LATEST = 0
+    RANDOM = 1
+    COMIC_ID = 2
+
+
+class XkcdNotFoundError(Exception):
+    pass
+
+
+class XkcdServerError(Exception):
+    pass
+
+
+def get_xkcd_bot_response(message):
+    original_content = message['content'].strip()
+    cropped = original_content[len('@xkcd '):]
+    command = cropped.strip()
+
+    xkcd_called_without_command = original_content == '@xkcd'
+
+    try:
+        if command == 'help' or xkcd_called_without_command:
+            return ("xkcd bot support these commands:"
+                    "\n* `@xkcd help` to show this help message."
+                    "\n* `@xkcd latest` fetch latest comic from xkcd."
```

Comment:
And [24](https://github.com/zulip/zulip/pull/2907/files#diff-4b7f0db39351c19065529a0000edb3caR24).